### PR TITLE
Adding notification

### PR DESCRIPTION
**Pull request by maximeroucher**

### CODE REVIEW

File: lib/service/local_notification_service.dart [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+    _localNotificationService.initialize(initializationSettings,
+        onDidReceiveNotificationResponse: onDidReceiveNotificationResponse,
+        onDidReceiveBackgroundNotificationResponse:
+            onDidReceiveBackgroundNotificationResponse);
+  }
+
+  Future<NotificationDetails> _notificationDetails() async {
+    AndroidNotificationDetails androidNotificationDetails =
+        const AndroidNotificationDetails("fr.myecl.titan", "TitanNotification",
+            channelDescription: "Notifications channel forr Titan",
```
Comment (Petitoto) [inline lib/service/local_notification_service.dart:44]:
```suggestion
            channelDescription: "Notifications channel for Titan",
```

File: .gitignore [hunk 2]
```diff
@@ -44,8 +44,15 @@ app.*.map.json
...
+# lock
+pubspec.lock
+
 # env 
-.env*
+.env
+
+# Firebase
+/android/app/google-services.json
+/ios/Runner/GoogleService-Info.plist
```
Comment (Petitoto) [inline .gitignore:55]: We should add a template for these files

Comment (maximeroucher) [reply .gitignore:55]: These files are entirely generated by Firebase, so I don't think we need to do so

**Review by Petitoto — APPROVED** [review]

File: lib/cinema/repositories/cinema_topic_repository.dart [hunk 3]
```diff
@@ -0,0 +1,23 @@
+import 'package:myecl/tools/repository/repository.dart';
+
+class CinemaTopicRepository extends Repository {
+  @override
+  // ignore: overridden_fields
+  final ext = 'notification/';
+  final prefix = "cinema_";
+
+  Future<bool> subscribeSession(String topic) async {
+    // return await create({}, suffix: "topics/$prefix$topic/subscribe");
```
Comment (Rotheem) [inline lib/cinema/repositories/cinema_topic_repository.dart:10]: Why is this line commented ?

File: lib/cinema/repositories/cinema_topic_repository.dart [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    return true;
+  }
+
+  Future<bool> unsubscribeSession(String topic) async {
+    // return await create({}, suffix: "topics/$prefix$topic/unsubscribe");
```
Comment (Rotheem) [inline lib/cinema/repositories/cinema_topic_repository.dart:15]: same

File: lib/cinema/repositories/session_repository.dart [hunk 5]
```diff
@@ -7,7 +7,9 @@ class SessionRepository extends Repository {
   final ext = 'cinema/sessions';
 
   Future<List<Session>> getAllSessions() async {
-    return (await getList()).map((e) => Session.fromJson(e)).toList();
+    final List<Session> sessionList =
```
Comment (Rotheem) [inline lib/cinema/repositories/session_repository.dart:10]: Why do you use a temporary variable ?

File: lib/cinema/ui/pages/main_page/session_card.dart [hunk 6]
```diff
@@ -133,24 +151,119 @@ class SessionCard extends HookConsumerWidget {
...
+                                                        context,
+                                                        TypeMsg.msg,
+                                                        "Le rappel a été supprimé");
+                                                  } else {
+                                                    localNotificationService.showNotification(
+                                                        Message(
+                                                            actionModule: '',
+                                                            actionTable: '',
+                                                            content:
+                                                                'La séance '
```
Comment (Rotheem) [inline lib/cinema/ui/pages/main_page/session_card.dart:205]: You should use text constants.

File: lib/cinema/ui/pages/main_page/session_card.dart [hunk 7]
```diff
@@ -133,24 +151,119 @@ class SessionCard extends HookConsumerWidget {
...
+                                                            deliveryDateTime: session
+                                                                .start
+                                                                .subtract(
+                                                                    const Duration(
+                                                                        minutes:
+                                                                            10))));
+                                                    displayToast(
+                                                        context,
+                                                        TypeMsg.msg,
+                                                        "Un rappel a été ajouté");
```
Comment (Rotheem) [inline lib/cinema/ui/pages/main_page/session_card.dart:220]: Same

File: lib/service/class/message.dart [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+  Map<String, dynamic> toJson() {
+    final data = <String, dynamic>{};
+    data['title'] = title;
+    data['content'] = content;
+    data['action_module'] = actionModule;
+    data['action_table'] = actionTable;
+    data['context'] = context;
+    data['is_visible'] = isVisible;
+    data['delivery_datetime'] = deliveryDateTime?.toIso8601String().split('T')[0];
```
Comment (Rotheem) [inline lib/service/class/message.dart:40]: Won't it need a change with PR #121 ?

File: lib/service/local_notification_service.dart [hunk 9]
```diff
@@ -0,0 +1,218 @@
+import 'dart:convert';
+
+import 'package:flutter_local_notifications/flutter_local_notifications.dart';
+import 'package:myecl/service/class/message.dart' as message_class;
+import 'package:myecl/service/provider_list.dart';
+import 'package:myecl/tools/functions.dart';
+import 'package:qlevar_router/qlevar_router.dart';
+import 'package:rxdart/subjects.dart';
+import 'package:timezone/timezone.dart' as tz;
+import 'package:timezone/data/latest.dart' as tz;
```
Comment (Rotheem) [inline lib/service/local_notification_service.dart:10]: It is strange to import two different packages under the same name

File: lib/service/local_notification_service.dart [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+            android: initializationSettingsAndroid,
+            iOS: initializationSettingsDarwin,
+            linux: initializationSettingsLinux);
+    _localNotificationService.initialize(initializationSettings,
+        onDidReceiveNotificationResponse: onDidReceiveNotificationResponse,
+        onDidReceiveBackgroundNotificationResponse:
+            onDidReceiveBackgroundNotificationResponse);
+  }
+
+  Future<NotificationDetails> _notificationDetails() async {
```
Comment (Rotheem) [inline lib/service/local_notification_service.dart:41]: Why is there a '_' only on this function ?

File: lib/service/local_notification_service.dart [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+    AndroidNotificationDetails androidNotificationDetails =
+        const AndroidNotificationDetails("fr.myecl.titan", "TitanNotification",
+            channelDescription: "Notifications channel for Titan",
+            importance: Importance.max,
```
Comment (Rotheem) [inline lib/service/local_notification_service.dart:45]: We will later need a way for user to change this

File: lib/service/local_notification_service.dart [hunk 12]
```diff
@@ -0,0 +1,218 @@
...
+  Future<String> handleAction(String actionModule, String actionTable) async {
+    final provider = providers[actionModule];
+    if (provider == null) {
+      return "";
+    }
+    final information = provider[actionTable];
+    if (information == null) {
+      return "";
+    }
+    final path = information.item1;
```
Comment (Rotheem) [inline lib/service/local_notification_service.dart:174]: Why don't you directly return information.item1 ?

File: lib/service/local_notification_service.dart [hunk 13]
```diff
@@ -0,0 +1,218 @@
...
+@pragma("vm:entry-point")
+void onDidReceiveBackgroundNotificationResponse(
+    NotificationResponse response) async {
+  if (response.payload == null) {
+    return;
+  }
+  final message = message_class.Message.fromJson(
+      jsonDecode(utf8.decode(response.payload!.runes.toList())));
+  if (message.actionModule != null && message.actionTable != null) {
+    final provider = providers[message.actionModule];
```
Comment (Rotheem) [inline lib/service/local_notification_service.dart:206]: This code is similar to handleAction, why not use this function ?

File: lib/cinema/repositories/cinema_topic_repository.dart (hunk 3 above)
Comment (maximeroucher) [reply lib/cinema/repositories/cinema_topic_repository.dart:10]: It was to mock an API response. This should not be commented anymore

File: lib/cinema/repositories/cinema_topic_repository.dart (hunk 4 above)
Comment (maximeroucher) [reply lib/cinema/repositories/cinema_topic_repository.dart:15]: Same :)

File: lib/cinema/repositories/session_repository.dart (hunk 5 above)
Comment (maximeroucher) [reply lib/cinema/repositories/session_repository.dart:10]: It was to print response.

File: lib/cinema/ui/pages/main_page/session_card.dart (hunk 6 above)
Comment (maximeroucher) [reply lib/cinema/ui/pages/main_page/session_card.dart:205]: This is a first iteration, the text might be updated

File: lib/service/class/message.dart (hunk 8 above)
Comment (maximeroucher) [reply lib/service/class/message.dart:40]: It first need to be merged

File: lib/service/local_notification_service.dart (hunk 10 above)
Comment (maximeroucher) [reply lib/service/local_notification_service.dart:41]: To declare it as private, I might rename it

File: lib/service/local_notification_service.dart (hunk 11 above)
Comment (maximeroucher) [reply lib/service/local_notification_service.dart:45]: Yes, maybe !

File: lib/service/local_notification_service.dart (hunk 12 above)
Comment (maximeroucher) [reply lib/service/local_notification_service.dart:174]: It was to print result

File: lib/service/local_notification_service.dart (hunk 13 above)
Comment (maximeroucher) [reply lib/service/local_notification_service.dart:206]: You are right !

File: lib/service/local_notification_service.dart (hunk 9 above)
Comment (maximeroucher) [reply lib/service/local_notification_service.dart:10]: They are both from the same module

**Review by Rotheem — APPROVED** [review]

**Pull request merged**